sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py

```diff
@@ -235,40 +259,19 @@ def test_remote_runner_display_data(self):
     p = Pipeline(remote_runner,
                  options=PipelineOptions(self.default_properties))
 
-    # TODO: Should not subclass ParDo. Switch to PTransform as soon as
-    # composite transforms support display data.
-    class SpecialParDo(beam.ParDo):
-      def __init__(self, fn, now):
-        super(SpecialParDo, self).__init__(fn)
-        self.fn = fn
-        self.now = now
-
-      # Make this a list to be accessible within closure
-      def display_data(self):
-        return {'asubcomponent': self.fn,
-                'a_class': SpecialParDo,
-                'a_time': self.now}
-
-    class SpecialDoFn(beam.DoFn):
-      def display_data(self):
-        return {'dofn_value': 42}
-
-      def process(self):
-        pass
-
     now = datetime.now()
     # pylint: disable=expression-not-assigned
     (p | ptransform.Create([1, 2, 3, 4, 5])
      | 'Do' >> SpecialParDo(SpecialDoFn(), now))
 
-    p.run()
+    # TODO(BEAM-366) Enable runner API on this test.
+    p.run(test_runner_api=False)
```

Comment:
Could you add the reason why? IIRC, it's because Runner API does a pickle-unpickle test which can crash under certain combinations of IDE/debugger settings and Python versions.

Reply:
As per BEAM-366, Runner API does not plumb through display_data for composite transforms. With this change we no longer have recursion errors. The recursion errors are caught in [1] and triggered the same codepath as when we set test_runner_api=False in [2], that's why the test was passing. However sometimes recursion error was not caught in [1], but was caught by Python IDE/debugger, and sometimes [3] Python 3.7 does not catch it. In these scenarios, the test used to fail.
[1] https://github.com/apache/beam/blob/c3c5999e3a82d865810f799767c179e2c17d304b/sdks/python/apache_beam/pipeline.py#L402
[2] https://github.com/apache/beam/blob/c3c5999e3a82d865810f799767c179e2c17d304b/sdks/python/apache_beam/pipeline.py#L599
[3] https://bugs.python.org/issue38593

Reply:
Overall, I think BEAM-366 should capture the underlying reason here.